public/gametrace.h

```diff
@@ -12,84 +12,439 @@
+	bool HasInteractsWithLayer( int nLayerIndex ) const		{ return ( m_nInteractsWith & ( 1ull << nLayerIndex ) ) != 0; }
+	bool HasInteractsExcludeLayer( int nLayerIndex ) const	{ return ( m_nInteractsExclude & ( 1ull << nLayerIndex ) ) != 0; }
+	void EnableInteractsAsLayer( int nLayer )				{ m_nInteractsAs |= ( 1ull << nLayer ); }
+	void EnableInteractsWithLayer( int nLayer )				{ m_nInteractsWith |= ( 1ull << nLayer ); }
+	void EnableInteractsExcludeLayer( int nLayer )			{ m_nInteractsExclude |= ( 1ull << nLayer ); }
+	void DisableInteractsAsLayer( int nLayer )				{ m_nInteractsAs &= ~( 1ull << nLayer ); }
+	void DisableInteractsWithLayer( int nLayer )			{ m_nInteractsWith &= ~( 1ull << nLayer ); }
+	void DisableInteractsExcludeLayer( int nLayer )			{ m_nInteractsExclude &= ~( 1ull << nLayer ); }
+	
+public:
+	// Which interaction layers do I interact or collide with? (e.g. I collide with LAYER_INDEX_CONTENTS_PASS_BULLETS because I am not a bullet)
+	// NOTE: This is analogous to the "solid mask" or "trace mask" in source 1 (bit mask of CONTENTS_* or 1<<LAYER_INDEX_*)
+	uint64 m_nInteractsWith;
+	
+	// Which interaction layers do I _not_ interact or collide with?  If my exclusion layers match m_nInteractsAs on the other object then no interaction happens.
+	uint64 m_nInteractsExclude;
+	
+	// Which interaction layers do I represent? (e.g. I am a LAYER_INDEX_CONTENTS_PLAYER_CLIP volume)
+	// NOTE: This is analogous to "contents" in source 1  (bit mask of CONTENTS_* or 1<<LAYER_INDEX_*)
+	uint64 m_nInteractsAs;
+	
+	uint32 m_nEntityIdsToIgnore[2]; // this is the ID of the game entity which should be ignored
+	uint32 m_nOwnerIdsToIgnore[2];	// this is the ID of the owner of the game entity which should be ignored
+	uint16 m_nHierarchyIds[2];		// this is an ID for the hierarchy of game entities (used to disable collision among objects in a hierarchy)
+	uint16 m_nObjectSetMask; 		// set of RnQueryObjectSet bits
+	uint8 m_nCollisionGroup; 		// one of the registered collision groups
+
+	bool m_bHitSolid : 1; 							// if true, then query will hit solid
+	bool m_bHitSolidRequiresGenerateContacts : 1; 	// if true, the FCOLLISION_FUNC_ENABLE_SOLID_CONTACT flag will be checked to hit solid
+	bool m_bHitTrigger : 1; 						// if true, then query will hit tirgger
+	bool m_bShouldIgnoreDisabledPairs : 1; 			// if true, then ignores if the query and shape entity IDs are in collision pairs
+	bool m_bIgnoreIfBothInteractWithHitboxes : 1;	// if true, then ignores if both query and shape interact with LAYER_INDEX_CONTENTS_HITBOX
+	bool m_bForceHitEverything : 1;					// if true, will hit any objects without any conditions
+	bool m_bUnknown : 1;							// haven't found where this is used yet
+};
+
+struct RnQueryAttr_t : public RnQueryShapeAttr_t
+{
+};
+
+class CTraceFilter : public RnQueryAttr_t
+{
+public:
+	CTraceFilter( int collisionGroup = COLLISION_GROUP_DEFAULT, bool bIterateEntities = true ) 

```

Comment:
I think it would be nice to provide few more constructors here to set other common members inline here, like providing entities to ignore, interaction mask etc. Altho I can see that filling up ``m_nHierarchyIds`` might be problematic since it's accessed within CBaseEntity, not sure if there are other means to access it.